ISOBMFF/source/MP4A.cpp

```diff
@@ -0,0 +1,196 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ ******************************************************************************/
+
+/*!
+ * @file        MP4A.cpp
+ * @copyright   (c) 2017, DigiDNA - www.digidna.net
+ * @author      https://github.com/leela9980
+ */
+
+#include <ISOBMFF/MP4A.hpp>
+#include <ISOBMFF/ContainerBox.hpp>
+
+namespace ISOBMFF
+{
+    class MP4A::IMPL
+    {
+        public:
+            
+            IMPL();
+            IMPL( const IMPL & o );
+            ~IMPL();
+            
+            uint16_t    _channelcount;
+            uint16_t    _samplesize;
+            uint32_t    _samplerate;
+            std::vector< std::shared_ptr< Box > > _boxes;
+    };  
+    
+    MP4A::MP4A():
+        FullBox( "mp4a" ),
+        impl( std::make_unique< IMPL >() )
+    {}
+    
+    MP4A::MP4A( const MP4A & o ):
+        FullBox( o ),
+        impl( std::make_unique< IMPL >( *( o.impl ) ) )
+    {}
+    
+    MP4A::MP4A( MP4A && o ) noexcept:
+        FullBox( std::move( o ) ),
+        impl( std::move( o.impl ) )
+    {
+        o.impl = nullptr;
+    }
+    
+    MP4A::~MP4A()
+    {}
+    
+    MP4A & MP4A::operator =( MP4A o )
+    {
+        FullBox::operator=( o );
+        swap( *( this ), o );
+        
+        return *( this );
+    }
+    
+    void swap( MP4A & o1, MP4A & o2 )
+    {
+        using std::swap;
+        
+        swap( static_cast< FullBox & >( o1 ), static_cast< FullBox & >( o2 ) );
+        swap( o1.impl, o2.impl );
+    }
+    
+    void MP4A::ReadData( Parser & parser, BinaryStream & stream )
+    {
+        ContainerBox container( "????" );
+        
+        // // reserved[5]
+        stream.ReadUInt8();
+        stream.ReadUInt8();
+        stream.ReadUInt8();
+        stream.ReadUInt8();
+        stream.ReadUInt8();
+        stream.ReadUInt8();
+
+        stream.ReadUInt16();
+
+        stream.ReadUInt32();
+        stream.ReadUInt32();
+
+        
+        SetChannelCount( stream.ReadBigEndianUInt16() );
+        SetSampleSize( stream.ReadBigEndianUInt16() );
+
+        // pre_defined1
+        stream.ReadUInt16();
+        // reserved1
+        stream.ReadUInt16();
+
+        uint32_t sampleRateFixed =  stream.ReadBigEndianUInt32();
```

Comment:
//      template unsigned int(32) samplerate = { default samplerate of media } << 16;